src/core/analysis-cache.ts

```diff
@@ -0,0 +1,106 @@
+/**
+ * @license
+ * Copyright (c) 2015 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {Document, ScannedDocument} from '../model/model';
+import {ParsedDocument} from '../parser/document';
+
+import {DependencyGraph} from './dependency-graph';
+
+export class AnalysisCache {
+  /**
+   * These are maps from resolved URLs to Promises of various stages of the
+   * analysis pipeline.
+   */
+
+  parsedDocumentPromises: Map<string, Promise<ParsedDocument<any, any>>>;
+  scannedDocumentPromises: Map<string, Promise<ScannedDocument>>;
+  analyzedDocumentPromises: Map<string, Promise<Document>>;
+
+  /**
+   * This is a map from a resolved url to a promise that will resolve when
+   * that document's dependencies have been scanned.
+   *
+   * We need to keep track of this separate from just the scanned document
```

Comment:
If we combined the "scan dependencies" and "resolve" steps then this cache wouldn't be necessary, but this PR already seems complex enough, and it seems better to punt it.